src/elasticsearch/index.ts

```diff
@@ -1,166 +1,440 @@
-  const rangeOperators = ['gt', 'lt', 'gte', 'lte', 'moreq', 'from', 'to']
+  protected customFilters: FiltersInterface = {}
+
+  /**
+   * Bodybuilder object
+   * – passed in from the outside so we don't need to include in this lib
+   */
+  protected queryChain: any
+
+  /**
+   * This is an empty clone of the original, empty bodybuilder query chain to use it as plain bodybuilder instance
+   */
+  protected emptyQueryChain: any
+
+  protected config: ElasticsearchQueryConfig
+
+  protected searchQuery: SearchQuery
 
-  // process applied filters
-  const appliedFilters = cloneDeep(searchQuery.getAppliedFilters()) // copy as function below modifies the object
-  if (appliedFilters.length > 0) {
-    let hasCatalogFilters = false
-    // apply default filters
+  protected optionsPrefix: string = '_options'
+
+  protected _hasCatalogFilters: boolean
+
+  /**
+   * @param {{ config: ElasticsearchQueryConfig, queryChain: any, searchQuery: SearchQuery, customFilters?: FiltersInterface }} 
+   */
+  constructor ({ config, queryChain, searchQuery, customFilters }: { config: ElasticsearchQueryConfig, queryChain: any, searchQuery: SearchQuery, customFilters?: FiltersInterface }) {
+    this.config = config
+    this.queryChain = queryChain
+    this.searchQuery = searchQuery
+
+    this.customFilters = Object.assign({}, this.customFilters, customFilters)
+
+    this.emptyQueryChain = queryChain.clone()
+  }
+  
+  /**
+   * @return {this}
+   */
+  public buildQueryBodyFromSearchQuery (): this {
+    if (this.searchQuery.getAppliedFilters().length > 0) {
+      /**
+       * @param {any} appliedFilters This is a copy of `searchQuery.getAppliedFilters()` because we going to modify the object
+       */
+      const appliedFilters = cloneDeep(this.searchQuery.getAppliedFilters())
+
+      this
+        .applyBaseFilters(appliedFilters)
+        .applyCatalogFilters(appliedFilters)
+        .applyAggregations()
+        .applyTextQuery()
+        .applySort()
+    }
+
+    return this
+  }
+
+  /**
+   * Apply all `default` scoped filters to `queryChain`
+   * @return {this}
+   */
+  protected applyBaseFilters (appliedFilters: AppliedFilter[]): this {
     appliedFilters.forEach(filter => {
-      if (checkIfObjectHasScope({ object: filter, scope: 'default' }) && Object.keys(filter.value).length) {
-        if (Object.keys(filter.value).every(v => (rangeOperators.indexOf(v) >= 0))) {
+      let filterApplied = false
+      let { value, attribute } = filter
+
+      if (!this.checkIfObjectHasScope({ object: filter, scope: 'default' })) {
+        return
+      }
+
+      value = typeof value === 'object' ? value : { 'in': [ value ] }
+      if (Object.keys(value).length > 0) {
+        attribute = this.getMapping(attribute)
+        const operator = Object.keys(value)[0]
+
+        this.getSortedFilters().forEach(filterHandler => {
+          // Prevent duplicate assign of filters, despite the fact you can't break `forEach`
```

Comment:
how about using `for in` loop? :thinking: then you can return from it and clear this `filterApplied`

Reply:
Done